Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp

```diff
@@ -4131,3 +4131,81 @@ WebKitFeatureList* webkit_settings_get_development_features(void)
 {
     return webkitFeatureListCreate(WebPreferences::internalDebugFeatures());
 }
+
+/**
+ * webkit_settings_apply_from_key_file:
+ * @settings: a #WebKitSettings
+ * @key_file: a #GKeyFile
+ * @group_name: Name of the group to read from @key_file
+ * @error: return location for error or %NULL to ignore
+ *
+ * Reads the contents of the given @group_name from the given @key_file and apply the value of
+ * each key/value to the corresponding property on the @settings.
+ *
+ * Returns: %TRUE if the settings were correctly applied or %FALSE otherwise.
+ *
+ * Since: 2.46
+ */
+gboolean webkit_settings_apply_from_key_file(WebKitSettings* settings, GKeyFile* keyFile, const gchar* groupName, GError ** error)
+{
+    g_return_val_if_fail(WEBKIT_IS_SETTINGS(settings), FALSE);
+    g_return_val_if_fail(keyFile, FALSE);
+
+    GObjectClass* klass = G_OBJECT_GET_CLASS(settings);
+    unsigned totalProperties = 0;
+    GUniquePtr<GParamSpec*> properties(g_object_class_list_properties(klass, &totalProperties));
+
+    if (!g_key_file_has_group(keyFile, groupName)) {
+        g_set_error(error, G_KEY_FILE_ERROR, G_KEY_FILE_ERROR_INVALID_VALUE, "The key file has no %s group", groupName);
+        return FALSE;
+    }
+
+    for (unsigned i = 0; i < totalProperties; i++) {
+        GUniqueOutPtr<GError> lookupError;
+        const char* propertyName = properties.get()[i]->name;
+        if (!g_key_file_has_key(keyFile, groupName, propertyName, &lookupError.outPtr())) {
+            if (lookupError) {
+                g_propagate_error(error, lookupError.release());
+                return FALSE;
+            }
+            continue; // Setting missing in GKeyFile, skip it.
+        }
+
+        switch (G_PARAM_SPEC_VALUE_TYPE(properties.get()[i])) {
+        case G_TYPE_BOOLEAN: {
+            gboolean value = g_key_file_get_boolean(keyFile, groupName, propertyName, &lookupError.outPtr());
+            if (!value && lookupError) {
+                g_propagate_error(error, lookupError.release());
+                return FALSE;
```

Comment:
If we want to make parse errors fatal, I think we should parse first and apply later, to make sure we either apply all setting or none. Also if we really want to apply settings on existing object, we should freeze notify before applying settings and thaw after.

Reply:
Good idea about freeze/thaw to avoid signal storms 👌🏼